#!/usr/bin/env python3
"""
MDDB Python Client Example

This example demonstrates how to use the MDDB gRPC client in Python.
"""

import grpc
from mddb_client import mddb_pb2, mddb_pb2_grpc


def main():
    # Connect to MDDB server
    channel = grpc.insecure_channel('localhost:11024')
    client = mddb_pb2_grpc.MDDBStub(channel)

    print("üîó Connected to MDDB server")
    print()

    # ========================================================================
    # Add a document
    # ========================================================================
    print("üìù Adding a document...")
    doc = client.Add(mddb_pb2.AddRequest(
        collection='blog',
        key='python-example',
        lang='en_US',
        meta={
            'category': mddb_pb2.MetaValues(values=['tutorial', 'python']),
            'author': mddb_pb2.MetaValues(values=['Python Developer']),
            'tags': mddb_pb2.MetaValues(values=['grpc', 'api', 'example'])
        },
        content_md='# Python gRPC Example\n\nThis document was created using the Python client!'
    ))
    print(f"‚úÖ Document added: {doc.id}")
    print(f"   Added at: {doc.added_at}")
    print()

    # ========================================================================
    # Get the document
    # ========================================================================
    print("üìñ Retrieving document...")
    doc = client.Get(mddb_pb2.GetRequest(
        collection='blog',
        key='python-example',
        lang='en_US',
        env={'year': '2024', 'language': 'Python'}
    ))
    print(f"‚úÖ Retrieved: {doc.key}")
    print(f"   Content: {doc.content_md[:50]}...")
    print()

    # ========================================================================
    # Search documents
    # ========================================================================
    print("üîç Searching for tutorial documents...")
    resp = client.Search(mddb_pb2.SearchRequest(
        collection='blog',
        filter_meta={
            'category': mddb_pb2.MetaValues(values=['tutorial'])
        },
        sort='updatedAt',
        asc=False,
        limit=10
    ))
    print(f"‚úÖ Found {len(resp.documents)} documents")
    for i, doc in enumerate(resp.documents, 1):
        print(f"   {i}. {doc.key} ({doc.lang})")
    print()

    # ========================================================================
    # Get server statistics
    # ========================================================================
    print("üìä Getting server statistics...")
    stats = client.Stats(mddb_pb2.StatsRequest())
    print(f"‚úÖ Server Stats:")
    print(f"   Database: {stats.database_path}")
    print(f"   Size: {stats.database_size / 1024:.2f} KB")
    print(f"   Mode: {stats.mode}")
    print(f"   Total Documents: {stats.total_documents}")
    print(f"   Total Revisions: {stats.total_revisions}")
    print()

    if stats.collections:
        print("   Collections:")
        for coll in stats.collections:
            print(f"     ‚Ä¢ {coll.name}: {coll.document_count} docs, "
                  f"{coll.revision_count} revisions")
    print()

    # ========================================================================
    # Create backup
    # ========================================================================
    print("üíæ Creating backup...")
    backup_resp = client.Backup(mddb_pb2.BackupRequest(
        to='python-backup.db'
    ))
    print(f"‚úÖ Backup created: {backup_resp.backup}")
    print()

    print("‚ú® All operations completed successfully!")
    channel.close()


if __name__ == '__main__':
    try:
        main()
    except grpc.RpcError as e:
        print(f"‚ùå gRPC Error: {e.code()} - {e.details()}")
    except Exception as e:
        print(f"‚ùå Error: {e}")
